File: smoke_tests/dart/insecure_random.dart
// Insecure Random Number Generation vulnerabilities in Dart

import 'dart:math';
import 'dart:convert';

// Test 1: Math.Random for security
String vulnerableMathRandom() {
  // VULNERABLE: Math.Random is not cryptographically secure
  var random = Random();
  return random.nextInt(1000000).toString().padLeft(6, '0');
}

// Test 2: Seeded with timestamp
String vulnerableTimestampSeed() {
  // VULNERABLE: Predictable seed
  var random = Random(DateTime.now().millisecondsSinceEpoch);
  return random.nextDouble().toString();
}

// Test 3: Static seed
String vulnerableStaticSeed() {
  // VULNERABLE: Same seed always
  var random = Random(42);
  return random.nextInt(1000000).toString();
}

// Test 4: Token from weak random
String vulnerableToken() {
  // VULNERABLE: Weak PRNG for token
  var random = Random();
  var chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  return List.generate(16, (_) => chars[random.nextInt(chars.length)]).join();
}

// Test 5: UUID-like with weak random
String vulnerableUuid() {
  // VULNERABLE: Weak random for UUID
  var random = Random();
  var bytes = List.generate(16, (_) => random.nextInt(256));
  return bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
}

// Test 6: Password generation with weak random
String vulnerablePassword() {
  // VULNERABLE: Password from weak PRNG
  var random = Random();
  var chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#\$%';
  return List.generate(12, (_) => chars[random.nextInt(chars.length)]).join();
}

// Test 7: OTP generation with weak random
String vulnerableOtp() {
  // VULNERABLE: OTP from weak PRNG
  var random = Random();
  return random.nextInt(1000000).toString().padLeft(6, '0');
}

// Test 8: Session ID with weak random
String vulnerableSessionId() {
  // VULNERABLE: Session ID from weak PRNG
  var random = Random();
  var bytes = List.generate(32, (_) => random.nextInt(256));
  return base64Encode(bytes);
}

// Test 9: Nonce with weak random
String vulnerableNonce() {
  // VULNERABLE: Nonce should be cryptographically random
  var random = Random();
  return random.nextInt(1 << 32).toRadixString(16);
}

// Test 10: IV generation with weak random
List<int> vulnerableIv() {
  // VULNERABLE: IV from weak PRNG
  var random = Random();
  return List.generate(16, (_) => random.nextInt(256));
}

// Test 11: Salt generation with weak random
List<int> vulnerableSalt() {
  // VULNERABLE: Salt from weak PRNG
  var random = Random();
  return List.generate(16, (_) => random.nextInt(256));
}

// Test 12: Shuffle with weak random
List<T> vulnerableShuffle<T>(List<T> list) {
  // VULNERABLE: Shuffle with weak PRNG
  var random = Random();
  var shuffled = List<T>.from(list);
  shuffled.shuffle(random);
  return shuffled;
}

// Test 13: Random selection with weak random
T vulnerableRandomSelect<T>(List<T> options) {
  // VULNERABLE: Selection with weak PRNG
  var random = Random();
  return options[random.nextInt(options.length)];
}

// Test 14: Verification code with weak random
String vulnerableVerificationCode() {
  // VULNERABLE: Verification code from weak PRNG
  var random = Random();
  return random.nextInt(10000).toString().padLeft(4, '0');
}

// Test 15: API key generation with weak random
String vulnerableApiKey() {
  // VULNERABLE: API key from weak PRNG
  var random = Random();
  var bytes = List.generate(24, (_) => random.nextInt(256));
  return base64Encode(bytes).replaceAll(RegExp(r'[+/=]'), '');
}

// SAFE: Using Random.secure()
String safeSecureRandom() {
  // SAFE: Cryptographically secure random
  var random = Random.secure();
  var bytes = List.generate(32, (_) => random.nextInt(256));
  return base64Encode(bytes);
}

File: smoke_tests/csharp/insecure_random.cs
// Insecure Randomness vulnerabilities in C#
using System;
using System.Security.Cryptography;
using System.Text;
using System.Web.Mvc;

namespace VulnerableApp
{
    public class RandomController : Controller
    {
        // Test 1: System.Random for security-sensitive token
        public ActionResult GenerateToken()
        {
            // VULNERABLE: System.Random is not cryptographically secure
            var random = new Random();
            var token = new StringBuilder();
            for (int i = 0; i < 32; i++)
            {
                token.Append(random.Next(16).ToString("x"));
            }
            return Content(token.ToString());
        }

        // Test 2: Random with predictable seed
        public ActionResult GenerateWithSeed()
        {
            // VULNERABLE: Predictable seed
            var random = new Random(42);
            return Json(random.Next());
        }

        // Test 3: Time-based seed
        public ActionResult TimeBasedRandom()
        {
            // VULNERABLE: Time-based seed is predictable
            var random = new Random((int)DateTime.Now.Ticks);
            return Json(random.Next());
        }

        // Test 4: Random for password generation
        public ActionResult GeneratePassword()
        {
            const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
            // VULNERABLE: Using Random for password
            var random = new Random();
            var password = new char[12];
            for (int i = 0; i < 12; i++)
            {
                password[i] = chars[random.Next(chars.Length)];
            }
            return Content(new string(password));
        }

        // Test 5: Random for session ID
        public ActionResult CreateSession()
        {
            // VULNERABLE: Session ID must be cryptographically random
            var random = new Random();
            string sessionId = random.Next().ToString("x8") +
                             random.Next().ToString("x8");
            return Content(sessionId);
        }

        // Test 6: Random for CSRF token
        public ActionResult GetCsrfToken()
        {
            // VULNERABLE: CSRF token needs crypto random
            var random = new Random();
            byte[] tokenBytes = new byte[16];
            for (int i = 0; i < 16; i++)
            {
                tokenBytes[i] = (byte)random.Next(256);
            }
            return Content(Convert.ToBase64String(tokenBytes));
        }

        // Test 7: Random for encryption key
        public ActionResult GenerateKey()
        {
            // VULNERABLE: Encryption keys must be crypto random
            var random = new Random();
            byte[] key = new byte[32];
            random.NextBytes(key);
            return Content(Convert.ToBase64String(key));
        }

        // Test 8: Random for IV/nonce
        public ActionResult GenerateIv()
        {
            // VULNERABLE: IV must be unpredictable
            var random = new Random();
            byte[] iv = new byte[16];
            random.NextBytes(iv);
            return Content(Convert.ToBase64String(iv));
        }

        // Test 9: Random for API key
        public ActionResult GenerateApiKey()
        {
            // VULNERABLE: API key should be crypto random
            var random = new Random();
            return Content(Guid.NewGuid().ToString() + random.Next().ToString("x8"));
        }

        // Test 10: Static Random instance (thread safety issues)
        private static Random _staticRandom = new Random();

        public ActionResult StaticRandom()
        {
            // VULNERABLE: Not thread-safe and not crypto secure
            return Json(_staticRandom.Next());
        }

        // Test 11: Random for OTP
        public ActionResult GenerateOtp()
        {
            // VULNERABLE: OTP must be crypto random
            var random = new Random();
            int otp = random.Next(100000, 999999);
            return Content(otp.ToString());
        }

        // Test 12: Environment.TickCount seed
        public ActionResult TickCountSeed()
        {
            // VULNERABLE: TickCount is predictable
            var random = new Random(Environment.TickCount);
            return Json(random.Next());
        }
    }
}
